server/server.go

```diff
@@ -1656,6 +1665,145 @@ func (s *StanServer) start(runningState State) error {
 	return nil
 }
 
+// startChannelGossiping sets up a NATS subscription to a cluster gossip
+// subject to receive messages from the cluster on what channels exist. It also
+// starts a goroutine which periodically publishes the channels this server
+// knows about. When a server discovers it's missing channels, it requests the
+// missing channels from another server and creates them. This should only be
+// called if the server is running in clustered mode.
+func (s *StanServer) startChannelGossiping() error {
+	// Setup sub for receiving channel lists for reconciliation.
+	if _, err := s.ncr.Subscribe(s.getChannelReconcileInbox(), s.reconcileChannels); err != nil {
+		return err
+	}
+
+	// Setup sub for receiving channel gossip.
+	gossipInbox := fmt.Sprintf("%s.%s.channels", defaultGossipPrefix, s.opts.ID)
+	if _, err := s.ncr.Subscribe(gossipInbox, s.processChannelGossip); err != nil {
+		return err
+	}
+
+	// Setup sub for receiving channel requests.
+	reqInbox := fmt.Sprintf("%s.%s.channels.request.%s", defaultGossipPrefix, s.opts.ID, s.opts.Clustering.NodeID)
+	if _, err := s.ncr.Subscribe(reqInbox, s.processChannelListRequest); err != nil {
+		return err
+	}
+
+	// Start publishing channel info.
+	go func() {
+		rand.Seed(time.Now().UnixNano())
+		for {
+			// Publish on a fixed interval with some random jitter.
+			randomDelay := time.Duration(rand.Intn(6)) * time.Second
+			select {
+			case <-time.After(s.opts.Clustering.GossipInterval + randomDelay):
+				s.channels.RLock()
+				count := uint32(len(s.channels.channels))
+				s.channels.RUnlock()
+				data, err := (&spb.ChannelGossip{NodeID: s.opts.Clustering.NodeID, NumChannels: count}).Marshal()
+				if err != nil {
+					panic(err)
+				}
+				s.ncr.PublishRequest(gossipInbox, reqInbox, data)
+			case <-s.shutdownCh:
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// reconcileChannels is a NATS handler that handles asynchronous responses from
+// a peer after requesting the list of channels. This is done asynchronously
+// since the list can exceed the max NATS message size, so we stream the list
+// in chunks asynchronously. Reconciling channels should be idempotent.
+func (s *StanServer) reconcileChannels(m *nats.Msg) {
+	// Message cannot be empty, we are supposed to receive
+	// a spb.CtrlMsg_Partitioning protocol.
+	if len(m.Data) == 0 {
+		return
+	}
+	req := spb.CtrlMsg{}
+	if err := req.Unmarshal(m.Data); err != nil {
+		s.log.Errorf("Error processing channel reconcile request: %v", err)
+		return
+	}
+	// Ignore messages from ourselves.
+	if req.ServerID == s.serverID {
+		return
+	}
+	channels, err := util.DecodeChannels(req.Data)
```

Comment:
It is understood that this is not the complete list. SendChannelLists() sends possible multiple NATS messages, so this current callback may be invoked multiple times for the same "list".

Reply:
Right, the server will send the complete list in chunks so this will get invoked for each chunks. It's possible this could get called redundantly if multiple servers are streaming their channel list, but this is OK since channel reconciliation is idempotent (and for now, we don't need to worry about channel deletion).